src/controllers/action-queue/push.js

```diff
@@ -0,0 +1,255 @@
+          const rawTx = asHex(executionTx.tx.signedTx)
+
+          const broadcastTx: BroadcastTx = { pluginId, rawTx }
+
+          return broadcastTx
+        })
+      )
+      const pushEvent: NewPushEvent = {
+        eventId,
+        broadcastTxs,
+        // Include pushMessage only for the last event because device should only wake up when the server finishes all push events.
+        pushMessage: index === dryrunOutputs.length - 1 ? pushMessage : undefined,
+        recurring: false,
+        trigger
+      }
+
+      return pushEvent
+    })
+  )
+
+  return pushEvents
+}
+
+export async function checkPushEvents(account: EdgeAccount, eventIds: string[]): Promise<boolean> {
+  const { rootLoginId: loginId } = account
+  const requestBody: PushRequestBody = {
+    apiKey: AIRBITZ_API_KEY,
+    deviceId,
+    loginId: asBase64(loginId)
+  }
+
+  const response = await fetch(`${pushServerUri}/v2/login`, {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json'
+    },
+    body: wasPushRequestBody(requestBody)
+  })
+
+  if (!response.ok) {
+    throw new Error(`Request failed with ${response.status}`)
+  }
+
+  const data = await response.json()
+  const loginPayload = asLoginPayload(data)
+  const eventStatusMap = loginPayload.events.reduce((map, eventStatus) => ({ ...map, [eventStatus.eventId]: eventStatus }), {})
+
+  const isEffective = eventIds.every(eventId => {
+    const status = eventStatusMap[eventId]
+    return status != null && ['triggered', 'complete'].includes(status.state)
+  })
+
+  return isEffective
+}
+
+export async function effectCanBeATrigger(account: EdgeAccount, effect: ActionEffect): Promise<boolean> {
+  try {
+    await actionEffectToPushTrigger(account, effect)
+  } catch (err) {
+    return false
+  }
+  return true
+}
+
+export async function uploadPushEvents(account: EdgeAccount, newPushEvents: NewPushEvent[]): Promise<void> {
+  const { rootLoginId: loginId } = account
+  const loginUpdatePayload: LoginUpdatePayload = {
+    createEvents: newPushEvents,
+    removeEvents: []
+  }
+  const requestBody: PushRequestBody = {
+    apiKey: AIRBITZ_API_KEY,
+    deviceId,
+    loginId: asBase64(loginId)
+  }
+  const response = await fetch(`${pushServerUri}/v2/login/update`, {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json'
+    },
+    body: wasPushRequestBody({ ...requestBody, data: wasLoginUpdatePayload(loginUpdatePayload) })
+  })
+
+  if (!response.ok) {
+    const responseBody = await response.text()
+    const responseData = asMaybe(asErrorResponse)(responseBody)
+    console.error(
+      'Failed push-server request:',
+      JSON.stringify(
+        {
+          responseBody,
+          responseData
+        },
+        null,
+        2
+      )
+    )
+    throw new Error(`Request failed with ${response.status}`)
+  }
+}
+
+async function actionEffectToPushTrigger(account: EdgeAccount, effect: ActionEffect): Promise<PushTrigger> {
```

Comment:
It might be simpler to return `| undefined`, and then use that to indicate the "not supported" case instead of throwing an exception. Then it's just `await actionEffectToPushTrigger(...) == null` instead of needing the `effectCanBeATrigger` wrapper (or make the wrapper simpler).

Reply:
I do like the self documenting aspect to the `effectCanBeATrigger` function. It's clearer what's going on at the call-site (one less comment to write). However, to make the implementation simpler we could return `undefined` which would remove the need for a try-catch, but we'd lose the loud error messages at runtime when a program attempts to call this function incorrectly (versus silently fail or fail later on).

Alternatively, error handling could be "cleaner" with a tuple type (this is a pattern in Go):

```ts
export async function effectCanBeATrigger(account: EdgeAccount, effect: ActionEffect): Promise<boolean> {
  const [, error] = await actionEffectToPushTrigger(account, effect)
  return error == null
```

Function utils could help with this:

```
const errant = (fn) => (...args) => { try { return [fn(...args)] } catch (e) { return [, e] } }
const errorless = (task) => task[1] == null
// ...
return errorless(await errant(actionEffectToPushTrigger(account, effect)))
```

Reply:
I don't see what's wrong with returning `undefined`. The type system forces you to deal with it right away, so there is no risk of silent failures. I don't mind the `effectCanBeATrigger` wrapper, as long as the internals are more sane.

Reply:
By returning `undefined` that means the call-site must implement it's own error handling for the error cases. Why would that be desirable over handling the error cases (mis-use of `actionEffectToPushTrigger`) once within the implementation of `actionEffectToPushTrigger`?

Reply:
I see, we should leverage the type system to force the usage to handle the undefined case.

Reply:
Fixed this.